pyingest/parsers/arxiv.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python
+
+# arxiv_dc: parser for arXiv records in Dublin Core XML format
+# M. Templeton, 2017 November 16
+
+import codecs
+from dubcore import DublinCoreParser
+
+class MissingAuthorException(Exception):
+    pass
+
+class MissingTitleException(Exception):
+    pass
+
+class MissingAbstractException(Exception):
+    pass
+
+class MissingIDException(Exception):
+    pass
+
+class MissingDateException(Exception):
+    pass
+
+class EmptyParserException(Exception):
+    pass
+
+class ArxivParser(DublinCoreParser):
+
+
+    def parse(self, fp, **kwargs):
+
+        arx = dict()
+
+        r = super(self.__class__, self).parse(fp, **kwargs)
+
+        if(len(r.keys()) == 0):
+            raise EmptyParserException("No dictionary.")
+
+        try:
+            arx['pubdate']  = r['dc:date'][-1]
+        except KeyError:
+            raise MissingDateException("Invalid record: no pubdate")
+        else:
+            pass
+
+        try:
+            arx['abstract'] = r['dc:description'][0]
+        except KeyError:
+            raise MissingAbstractException("Invalid record: no abstract")
+        else:
+            pass
+
+        try:
+            arx['title']    = [r['dc:title'][-1]]
+        except KeyError:
+            raise MissingTitleException("Invalid record: no title")
+        else:
+            pass
+
+        try:
+            arx['author']  = r['dc:creator']
+        except KeyError:
+            raise MissingAuthorException("Invalid record: no author(s)")
+        else:
+            pass
+
+        try:
+            arx['keyword'] = r['dc:subject']
+        except KeyError:
+            raise MissingAbstractException("Invalid record: no subjects")
+        else:
+            pass
+
+        try:
+            make_extras(r['dc:identifier'])
+        except KeyError:
+            raise MissingIDException("Invalid record: no identifier")
+        else:
+            testdoi,url = make_extras(r['dc:identifier'])
+            if testdoi != [None]:
+                arx['doi'] = testdoi
+
+        try:
+            arx['bibcode'] = make_bibcode(url,arx['author'])
+        except KeyError:
+            raise MissingBibcodeException("Invalid record: cant generate bibcode")
+        else:
+            pass
+
+        return arx
+
+
+def make_extras(ids):
+    doi = None
+    url = None
+    for x in ids:
+        if u'doi:' in x:
+            doi = x
+        if u'http' in x:
+            if u'arxiv.org' in x:
+                url = x
+
+    return [doi],url
+
+
+def make_bibcode(url,authors):
```

Comment:
Paragraph 1: Yes I did.  I seem to remember encountering some error at some point doing that.  But yes, I agree with what you're saying, and I will try to track down and handle whatever error I got using the ID.

Paragraph 2: This is a good point, and it's why I'd like to meet on Monday.  When we last talked about the specs for this, I got the impression that the bibcode didn't have to rigorously match what the classic parser gave because they'd be reconciled.  But in light of discussion at the Jan 26 scrum, I wonder if this is not true?  Regardless, I will implement the existing bibcode generation as exactly as I can if it is needed.